File: app/src/main/java/de/christinecoenen/code/zapp/app/settings/ui/ChannelSelectionActivity.java
package de.christinecoenen.code.zapp.app.settings.ui;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.RecyclerView;

import com.woxthebox.draglistview.DragListView;

import de.christinecoenen.code.zapp.R;
import de.christinecoenen.code.zapp.databinding.ActivityChannelSelectionBinding;
import de.christinecoenen.code.zapp.models.ISortableChannelList;
import de.christinecoenen.code.zapp.models.json.SortableJsonChannelList;
import de.christinecoenen.code.zapp.utils.view.GridAutofitLayoutManager;
import de.christinecoenen.code.zapp.utils.view.SimpleDragListListener;

public class ChannelSelectionActivity extends AppCompatActivity {

	private ISortableChannelList channelList;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		ActivityChannelSelectionBinding binding = ActivityChannelSelectionBinding.inflate(getLayoutInflater());
		setContentView(binding.getRoot());

		DragListView channelListView = binding.draglistChannelSelection;

		// adapter
		channelList = new SortableJsonChannelList(this);
		final ChannelSelectionAdapter listAdapter = new ChannelSelectionAdapter(this);
		listAdapter.setItemList(channelList.getList());

		// view
		RecyclerView.LayoutManager layoutManager = new GridAutofitLayoutManager(this, 120);
		channelListView.setLayoutManager(layoutManager);
		channelListView.setAdapter(listAdapter, true);
		channelListView.getRecyclerView().setVerticalScrollBarEnabled(true);

		channelListView.setDragListListener(new SimpleDragListListener() {
			@Override
			public void onItemDragEnded(int fromPosition, int toPosition) {
				if (fromPosition != toPosition) {
					channelList.persistChannelOrder();
				}
			}
		});
	}

	@Override
	protected void onPause() {
		super.onPause();
		channelList.persistChannelOrder();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.activity_channel_selection, menu);
		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		switch (item.getItemId()) {
			case R.id.menu_help:
				openHelpDialog();
				return true;
		}

		return super.onOptionsItemSelected(item);
	}

	private void openHelpDialog() {
		DialogFragment newFragment = new ChannelSelectionHelpDialog();
		newFragment.show(getSupportFragmentManager(), "help");
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/settings/ui/ChannelSelectionActivity.kt
package de.christinecoenen.code.zapp.app.settings.ui

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.databinding.ActivityChannelSelectionBinding
import de.christinecoenen.code.zapp.models.channels.ISortableChannelList
import de.christinecoenen.code.zapp.models.channels.json.SortableJsonChannelList
import de.christinecoenen.code.zapp.utils.view.GridAutofitLayoutManager
import de.christinecoenen.code.zapp.utils.view.SimpleDragListListener

class ChannelSelectionActivity : AppCompatActivity() {

	private lateinit var channelList: ISortableChannelList

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		val binding = ActivityChannelSelectionBinding.inflate(layoutInflater)
		setContentView(binding.root)

		// adapter
		channelList = SortableJsonChannelList(this)

		val listAdapter = ChannelSelectionAdapter(this)
		listAdapter.itemList = channelList.list

		// view
		val layoutManager = GridAutofitLayoutManager(this, 120)

		binding.draglistChannelSelection.apply {

			setLayoutManager(layoutManager)
			setAdapter(listAdapter, true)
			recyclerView.isVerticalScrollBarEnabled = true

			setDragListListener(object : SimpleDragListListener() {
				override fun onItemDragEnded(fromPosition: Int, toPosition: Int) {
					if (fromPosition != toPosition) {
						channelList.persistChannelOrder()
					}
				}
			})
		}
	}

	override fun onPause() {
		super.onPause()

		channelList.persistChannelOrder()
	}

	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		menuInflater.inflate(R.menu.activity_channel_selection, menu)
		return super.onCreateOptionsMenu(menu)
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		return when (item.itemId) {
			R.id.menu_help -> {
				openHelpDialog()
				true
			}
			else -> {
				super.onOptionsItemSelected(item)
			}
		}
	}

	private fun openHelpDialog() {
		ChannelSelectionHelpDialog().apply {
			show(supportFragmentManager, "help")
		}
	}
}
